File: ts/src/compressors/verbatim.ts
/**
 * Verbatim — the no-op compressor.
 *
 * Used when content must not be altered: code, diffs, stack traces, tool
 * schemas, short messages that wouldn't compress well. Also the safe default
 * when the classifier emits UNKNOWN.
 *
 * Returns input unchanged with accurate input/output token counts so
 * telemetry still reports meaningful totals (ratio is always 1.0).
 */

import { makeStats } from "../stats.js";
import { countMessageTokens } from "../tokens.js";
import type { ChatMessage } from "../types.js";
import type { Compressor, CompressResult } from "./base.js";

/** A pass-through Compressor that preserves messages exactly. */
export class Verbatim implements Compressor {
    readonly name = "verbatim";

    compress(messages: ChatMessage[]): CompressResult {
        const tokens = countMessageTokens(messages);
        return [
            messages,
            makeStats({
                inputTokens: tokens,
                outputTokens: tokens,
                ratio: 1.0,
                method: "verbatim",
            }),
        ];
    }

    async compressAsync(messages: ChatMessage[]): Promise<CompressResult> {
        return this.compress(messages);
    }
}

File: rust/src/compressors/verbatim.rs
//! Verbatim — the no-op compressor. Returns input unchanged with accurate
//! token counts so telemetry stays meaningful (ratio always 1.0).

use crate::compressors::Compressor;
use crate::json::Value;
use crate::stats::CompressionStats;
use crate::tokens::count_message_tokens;

#[derive(Default)]
pub struct Verbatim;

impl Compressor for Verbatim {
    fn name(&self) -> &'static str {
        "verbatim"
    }

    fn compress(&self, messages: &[Value]) -> (Vec<Value>, CompressionStats) {
        let tokens = count_message_tokens(messages);
        (
            messages.to_vec(),
            CompressionStats::counted("verbatim", tokens, tokens),
        )
    }
}
